node/impl/full/eth.go

```diff
@@ -1027,6 +1028,126 @@ func (a *EthModule) EthTraceTransaction(ctx context.Context, txHash string) ([]*
 	return txTraces, nil
 }
 
+func (a *EthModule) EthBlockNumberFromString(ctx context.Context, block string) (ethtypes.EthUint64, error) {
+
+	head := a.Chain.GetHeaviestTipSet()
+	switch block {
+	case "earliest":
+		return 0, xerrors.Errorf("block param \"earliest\" is not supported")
+	case "pending":
+		return ethtypes.EthUint64(head.Height()), nil
+	case "latest":
+		parent, err := a.Chain.GetTipSetFromKey(ctx, head.Parents())
+		if err != nil {
+			return 0, fmt.Errorf("cannot get parent tipset")
+		}
+		return ethtypes.EthUint64(parent.Height()), nil
+	case "safe":
+		latestHeight := head.Height() - 1
+		safeHeight := latestHeight - ethtypes.SafeEpochDelay
+		return ethtypes.EthUint64(safeHeight), nil
+	default:
+		blockNum, err := ethtypes.EthUint64FromHex(block)
+		if err != nil {
+			return 0, xerrors.Errorf("cannot parse fromBlock: %w", err)
+		}
+		return blockNum, err
+	}
+}
+
+func (a *EthModule) EthTraceFilter(ctx context.Context, filter ethtypes.EthTraceFilterCriteria) ([]*ethtypes.EthTraceFilterResult, error) {
+	fromBlock, err := a.EthBlockNumberFromString(ctx, filter.FromBlock)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot parse fromBlock: %w", err)
+	}
+
+	toBlock, err := a.EthBlockNumberFromString(ctx, filter.ToBlock)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot parse toBlock: %w", err)
+	}
+
+	var results []*ethtypes.EthTraceFilterResult
+	traceCounter := 0
+
+	for blkNum := fromBlock; blkNum <= toBlock; blkNum++ {
+		blockTraces, err := a.EthTraceBlock(ctx, strconv.FormatUint(uint64(blkNum), 10))
+		if err != nil {
+			return nil, xerrors.Errorf("cannot get trace for block %d: %w", blkNum, err)
+		}
+
+		for _, blockTrace := range blockTraces {
+			if matchFilterCriteria(blockTrace, filter) {
+				traceCounter++
+				if traceCounter <= filter.After {
+					continue
+				}
+
+				txTrace := ethtypes.EthTraceFilterResult{
+					EthTrace:            blockTrace.EthTrace,
+					BlockHash:           blockTrace.BlockHash,
+					BlockNumber:         blockTrace.BlockNumber,
+					TransactionHash:     blockTrace.TransactionHash,
+					TransactionPosition: blockTrace.TransactionPosition,
+				}
+				results = append(results, &txTrace)
+
+				// If Count is specified, limit the results
+				if filter.Count > 0 && len(results) >= filter.Count {
```

Comment:
I think we're going to need a maximum count for this because this could allow users to fetch unreasonably large results.

You can see an example of this in the `Events.MaxFilterResults` configuration option and how it gets used in the various components that need it (grep for `MaxFilterResults`); but the thing we shouldn't copy from that is that it just truncates your results silently: https://github.com/filecoin-project/lotus/blob/1e8bc10b04ce65554fc9b5e8c289dec7b57daf83/chain/events/filter/event.go#L141-L144 - I would much rather than API return an error so I know I'm not missing out on information so I can tighten my filter.

It would be great to find out what an ethereum implementation uses for a max, and what they do when they hit max, and try and copy that.

Reply:
I think this is being misinterpreted a bit. My understanding is that `after` and `count` work similarly to `LIMIT` and `OFFSET` in SQL.

- `count` specifies the maximum number of results to return.
- `after` skips a specified number of results before starting to return data.

Together, they provide fine-grained control over result pagination. I don't think they need to create errors.

Reply:
I think a limit is still needed because we can't have this API allocating and trying to return _massive_ slices. If you specify a wide enough block range and wide enough filter then `results` could become unreasonably large. Regardless of what `Count` and `After` is specified, there needs to be some limit on `len(results)` to force the consumer to opt in to using `Count` and `After` properly for paginated results. I don't know what that limit is, perhaps it starts the same as the events limit of 10,000. But for this API, unlike the events API, I think returning an error when you hit that limit might be best because then at least you know you've maxxed out and need to try an alternative strategy.